### PR TITLE
Update: add support for parser (fixes #17)

### LINKED ISSUE #17: Add parser option
Since ESLint v0.14.0 the parser to use can be configured. But eslint-tester does not provide any option to pass a different parser during the tests, which makes difficult for us to test rules on any other parser than espree.

### PR DESCRIPTION
**Pull request by yannickcr**

### CODE REVIEW

File: tests/lib/eslint-tester.js [hunk 1]
```diff
@@ -329,6 +329,27 @@ describe("ESLintTester", function() {
         });
     });
 
+    it("should pass-through the parser to the rule", function() {
```
Comment (nzakas) [inline tests/lib/eslint-tester.js]: It's probably best to include one test with a custom parser (maybe just Esprima) to make sure you can get something other than error message.

Comment (yannickcr) [inline tests/lib/eslint-tester.js]: I did not wanted to add a new devDependency just for one test, but if you are ok this that can do it.

Also, how can I test inside the rule that we are using the right parser ? I think the parser is not exposed to the rule, is it ?

Comment (nzakas) [inline tests/lib/eslint-tester.js]: dev dependencies are cheap, feel free to add. 

It's probably easiest to expose the parser being used on ESLintTester. Maybe `getParser()`? And then compare it to what you get from `require()`

Comment (yannickcr) [inline tests/lib/eslint-tester.js]: Ok for the devDependency.

Just to be sure, you're suggesting to add a `getParser()` method on `context` in ESLint ?

Comment (ilyavolodin) [inline tests/lib/eslint-tester.js]: To verify that parser is working, you can probably just use Esprima and pass it some ES6 that it doesn't yet support. Or JSX, and make sure it fails. I don't think exposing parser within context is a good idea.

Comment (nzakas) [inline tests/lib/eslint-tester.js]: @yannickcr No, I'm saying add `getParser()` on ESLintTester.

@ilyavolodin I'm not comfortable with that type of testing - you're relying on implicit failures to determine something that should be explicit. We should verify that the parsing succeeds, and also verify that the name of the parser was used as expected.

**Comment by yannickcr** [comment]
Tried something, let me know if that's what you were thinking.

File: lib/eslint-tester.js [hunk 2]
```diff
@@ -11,11 +11,11 @@
  * eslintTester.add("{ruleName}", {
  *      valid: [
  *          "{code}",
- *          { code: "{code}", args: {args}, global: {globals}, globals: {globals}, settings: {settings} }
+ *          { code: "{code}", args: {args}, global: {globals}, globals: {globals}, parser: {parser}, settings: {settings} }
  *      ],
  *      invalid: [
  *          { code: "{code}", errors: {numErrors} },
- *          { code: "{code}", args: {args}, global: {globals}, settings: {settings} errors: [{ message: "{errorMessage}", type: "{errorNodeType}"}] }
+ *          { code: "{code}", args: {args}, global: {globals}, parser: {parser}, settings: {settings}, errors: [{ message: "{errorMessage}", type: "{errorNodeType}"}] }
```
Comment (nzakas) [inline lib/eslint-tester.js]: Parser is just a string, right? Should have quotes.

File: lib/eslint-tester.js [hunk 3]
```diff
@@ -181,7 +188,16 @@ ESLintTester.prototype = {
         });
 
         return result.suite;
+    },
+
+    /**
+     * Gets the name of the last used parser.
+     * @returns {string} The name of the last used parser.
+     */
+    getParser: function() {
```
Comment (nzakas) [inline lib/eslint-tester.js]: Just realized this doesn't make sense because each test can have a different parser. :-/ Need to think about this some more.

Comment (yannickcr) [inline lib/eslint-tester.js]: Yes, so we could only know which parser was used last.

Also, I'm getting the information from the configuration and not from ESLint so the test is not really great, but ESLint does not expose any information about the used parser.

Comment (nzakas) [inline lib/eslint-tester.js]: OK, got it. All we really need to test is that the `parser` property is being set on the config that is passed to `eslint.verify()`. So you can use a Sinon spy to ensure that: http://sinonjs.org/docs/#spies

Comment (yannickcr) [inline lib/eslint-tester.js]: Good idea! I'll do this.

**Comment by Cellule** [comment]
Would you be open to allow setting a default parser.
For instance, you could do `var eslintTester = new ESLintTester(eslint, {parser: "Esprima"});` and then the default config for a test would include this parser avoiding the need to declaring it for every test case.

Also, it would be nice if we could highjack completely the default parser. I know this is radical, but it would allow running every test in eslint with another parser. We could do something like
```js
var ESLintTester = require("eslint-tester");
// Highjack
ESLintTester.defaultParser = "Esprima";
```
Then every subsequent test would use this parser if not specified otherwise.

**Comment by nzakas** [comment]
@Cellule please file an issue so we can talk about your idea.

**Comment by yannickcr** [comment]
@nzakas Did you had time to look at the last revision with sinon.js ?

**Comment by nzakas** [comment]
Oops, completely missed that you had updated it. This looks good to me. @ilyavolodin what do you think?

**Comment by ilyavolodin** [comment]
LGTM. Should be ready to merge.

**Pull request merged**